### DISCUSSION TITLE
Idea: collaborate more tightly with Git for Windows to provide a working `mingw-w64-git`

### DISCUSSION THREAD

**Discussion by dscho** [msys2/MINGW-packages#16383] [Ideas]
The main obstacle when I migrated Git for Windows from MSys to MSYS2 was that the MSYS2 runtime is opinionated in a way that disagrees with Git's expectations (as documented in its test suite).

That is the reason why Git for Windows ships with a fork of the MSYS2 runtime, and that in turn is the obstacle why the Git for Windows cannot trivially maintain a `mingw-w64-git` definition in `msys2/MINGW-packages`: it would not pass the test suite.

So here is an idea: try to work out how the expectations of Git can be met, maybe by making minor adjustments to the MSYS2 runtime, and once that is done, modify [Git for Windows' `mingw-w64-git` package](https://github.com/git-for-windows/MINGW-packages/tree/HEAD/mingw-w64-git) in a way that the identical definition can be used in `msys2/MINGW-packages` (one thing that would obviously need to be changed is that `mingw-w64-git` _must not_ put a `git-bash.exe` into the root directory 😁).

Are there people interested in working with me on this project?

EDIT: Here is an analysis of the deviations between MSYS2's and Git for Windows' flavor of the MSYS2 runtime:

- [x] git-for-windows/msys2-runtime@589fac65ff: this is a real bug fix for running `find` on a FAT file system and should probably go into MSYS2 as-is
- [x] git-for-windows/msys2-runtime@3e9c003101: this is a real bug fix for setting empty-valued environment variables (think about the difference between `${var:-hello}` and `${var-hello}` after setting `export var=`). The fix might even be welcome in the Cygwin project. -> https://github.com/msys2/msys2-runtime/pull/101
- [ ] git-for-windows/msys2-runtime@1e0136ee86: this might be a big controversial, but Git's test suite needs this (and thereby suggests that this is standard Linux behavior, something Cygwin always puts above POSIX)
- [x] git-for-windows/msys2-runtime@0b2d287629: a real bug fix that should be contributed to msys2/msys2-runtime -> https://github.com/msys2/msys2-runtime/pull/143 https://github.com/msys2/msys2-tests/pull/33
- [x] git-for-windows/msys2-runtime@a5012b9e83: a documentation improvement that should be contributed to the Cygwin project 
- [x] git-for-windows/msys2-runtime@d2456ea67e: improving debug facilities in the MSYS2 runtime that should go into msys2/msys2-runtime -> https://github.com/msys2/msys2-runtime/pull/148
- [x] git-for-windows/msys2-runtime@78fbdbbd5e: readibility (and _slight_ performance) improvement that should go into msys2/msys2-runtime -> https://github.com/msys2/msys2-runtime/pull/149
- [x] git-for-windows/msys2-runtime@707f9b218a, git-for-windows/msys2-runtime@c312f3d3c6, git-for-windows/msys2-runtime@99a84e0081, git-for-windows/msys2-runtime@6efd127eeb, git-for-windows/msys2-runtime@4076235378, git-for-windows/msys2-runtime@12461acc8a, git-for-windows/msys2-runtime@aeb2e423d9, git-for-windows/msys2-runtime@9f3e592333, and git-for-windows/msys2-runtime@92a0032d1a: adjust the implicit Windows <-> Unix-y path conversion to accommodate common Git invocations better, i.e. accommodating expectations documented by Git's test suite.
- [x] git-for-windows/msys2-runtime@f874ac108a: in addition to the horrible-to-use, poorly-named `MSYS2_ENV_CONV_EXCL` and `MSYS2_ARG_CONV_EXCL` variables, this commit introduces a poorly-named but at least dirt-easy-to-use `MSYS_NO_PATHCONV` variable to stop the MSYS2 runtime from misinterpreting command-line arguments starting with a slash as referring to an absolute Unix-y path that needs to be converted to a valid Windows one (think `cmd /?`).
- [x] git-for-windows/msys2-runtime@ad711d9aa5, git-for-windows/msys2-runtime@90d8aad19b and git-for-windows/msys2-runtime@4935495078: offer to use the standard Windows profile directories as home directories (as is expected by Git for Windows). [I contributed two iterations of these patches](https://inbox.sourceware.org/cygwin-patches/0Lg1Tn-1YnzUw0ScN-00pcgi@mail.gmx.com/T/#t) to the Cygwin project already, and simply have to find the time to do the final touch-up. It's on my ever-growing TODO list.
- [ ] git-for-windows/msys2-runtime@07e3f3f107: this is a net improvement that should go into msys2/msys2-runtime as-is.
- [ ] git-for-windows/msys2-runtime@99c3c01445: this should probably go to the Cygwin project, making it harder to introduce Unicode conversions that inadvertently use the private code page intended to pretend that Windows has no forbidden filename characters
- [ ] git-for-windows/msys2-runtime@d6849e0baa: this fixes a real bug, and I am not quite certain whether this should go into Cygwin or just msys2/msys2-runtime
- [ ] git-for-windows/msys2-runtime@1fbb6ee0c1: needs analysis whether this should directly go into msys2/msys2-runtime, or into Cygwin
- [x] git-for-windows/msys2-runtime@65ce633af3 (with git-for-windows/msys2-runtime@f5f594c2ff squashed in): this works around (as well as possible) issues calling an MSYS2 runtime from another MSYS2 runtime (or even the Cygwin runtime); It is not possible to reuse the pseudo terminals, but at least most other operations should work.
- [x] git-for-windows/msys2-runtime@bec3d608e8: this is a net improvement: it lets `uname` report the commit from which the MSYS2 runtime was built. This should go into msys2/msys2-runtime; We may eventually even want to contribute this to the Cygwin project, they may like it, too.

There are a couple of challenges with this idea outside of the MSYS2 runtime:
- [x] Git for Windows' `mingw-w64-git` installs a `/git-bash.exe` (and other files that are most inappropriate for MSYS2). We will need to split them into their own package, e.g. `mingw-w64-git-for-windows-root`, that only gets built and installed in the Git for Windows project.
- [x] Several of MSYS2's defaults are not amenable to Git for Windows and have to be overridden in the latter, e.g. to use the Windows profile directory as home directory rather than `/home/<user>` (which would be unwritable in an installed Git for Windows). There is really no good way to override this other than editing files that are owned by MSYS2's packages (such as `filesystem`). Git for Windows does this via a hack: [the post-install script of the `mingw-w64-git-extra` package](https://github.com/git-for-windows/build-extra/blob/HEAD/git-extra/git-extra.install) performs those edits as needed. It would be nice if we found a non-hacky solution for these modifications.
- [x] Git for Windows builds a libcurl that offers both Secure Channel and OpenSSL backends (but not GNU TLS) with a runtime switch (`CURL_SSL_BACKEND` for `curl.exe`, `http.sslBackend` for Git). The MSYS2 project insteads builds various flavors of `libcurl`; The user has to decide which flavor to install and then cannot switch between them at runtime. [I am ready to bite the bullet](https://github.com/git-for-windows/git/issues/4350#issuecomment-1485041503) and adjust Git for Windows' libcurl to align with MSYS2's paradigm (and I think I found a hacky way to still sort of support runtime-switching).

**Comment by DakshKK** [comment]
Interested, but I lack sufficient knowledge to do anything beneficial or fundamental.

**Comment by lazka** [comment]
Thanks @dscho. We had a small meetup just now and talked about working through your list.

Some small notes that came out of it (just from me and @elieux):

* we are generally fine with small breakages for MSYS2 users due to behavior differences, as we think this is more important for the project health long term
* upstreaming to cygwin will come later, to keep this doable

**Comment by Biswa96** [comment]
> * Several of MSYS2's defaults are not amenable to Git for Windows and have to be overridden in the latter, e.g. to use the Windows profile directory as home directory rather than `/home/<user>`

Is it possible not to change that behavior please? 🥺 I prefer config files to be store in `/home/$USER` directory. Alternatively, add option or config to restore that default directory.

**Reply by lazka** [reply]
That's definitely not something we will change by default.

**Reply by dscho** [reply]
Right, I should have clarified that my plan was _never_ to force Git for Windows' behavior onto MSYS2 here. Instead, I would like to find ways how we can keep the behaviors separate _without_ requiring Git for Windows to edit `filesystems`' version of `/etc/nsswitch.conf`.

**Comment by lazka** [comment]
I've started with something easy: https://github.com/msys2/msys2-runtime/pull/143

How do we want to keep track of the list above? Can I edit the original post, check those boxes and add links to PRs?

**Reply by rimrul** [reply]
> How do we want to keep track of the list above? Can I edit the original post, check those boxes and add links to PRs?

That's probably the cleanest approach.

**Reply by rimrul** [reply]
> https://github.com/git-for-windows/msys2-runtime/commit/ad711d9aa5, https://github.com/git-for-windows/msys2-runtime/commit/90d8aad19b and https://github.com/git-for-windows/msys2-runtime/commit/4935495078: offer to use the standard Windows profile directories as home directories (as is expected by Git for Windows). [I contributed two iterations of these patches](https://inbox.sourceware.org/cygwin-patches/0Lg1Tn-1YnzUw0ScN-00pcgi@mail.gmx.com/T/#t) to the Cygwin project already, and simply have to find the time to do the final touch-up. It's on my ever-growing TODO list.

The first of these three patches made it to cygwin as https://github.com/cygwin/cygwin/commit/27376c60a9b870e7ba456551165897a7408b5886

IDK how to keep track of that though. Can't exactly partially check a checkbox.

**Reply by dscho** [reply]
> The first of these three patches made it to cygwin as [cygwin/cygwin@27376c6](https://github.com/cygwin/cygwin/commit/27376c60a9b870e7ba456551165897a7408b5886)
> 
> IDK how to keep track of that though. Can't exactly partiallycheck a checkbox.

Right. I am working on the second and third patch for that, and then we can check the box.

**Reply by dscho** [reply]
So the second of these three patches made it to cygwin as https://github.com/cygwin/cygwin/commit/cc8404039f50c36100fbad85a3af22d579d3576a. Unfortunately I had to rewrite the third patch completely, and add a bug fix for the first.

I saw that @lazka opened https://github.com/msys2/msys2-runtime/pull/148 and https://github.com/msys2/msys2-runtime/pull/149... Thank you so much!

**Comment by rimrul** [comment]
> Git for Windows builds a libcurl that offers both Secure Channel and OpenSSL backends (but not GNU TLS) with a runtime switch (CURL_SSL_BACKEND for curl.exe, http.sslBackend for Git). The MSYS2 project insteads builds various flavors of libcurl; The user has to decide which flavor to install and then cannot switch between them at runtime. https://github.com/git-for-windows/git/issues/4350#issuecomment-1485041503 and adjust Git for Windows' libcurl to align with MSYS2's paradigm (and I think I found a hacky way to still sort of support runtime-switching).

This is done (https://github.com/git-for-windows/MINGW-packages/pull/76)

> https://github.com/git-for-windows/msys2-runtime/commit/f874ac108a: in addition to the horrible-to-use, poorly-named MSYS2_ENV_CONV_EXCL and MSYS2_ARG_CONV_EXCL variables, this commit introduces a poorly-named but at least dirt-easy-to-use MSYS_NO_PATHCONV variable to stop the MSYS2 runtime from misinterpreting command-line arguments starting with a slash as referring to an absolute Unix-y path that needs to be converted to a valid Windows one (think cmd /?).

This is done (https://github.com/msys2/msys2-runtime/pull/181)

> https://github.com/git-for-windows/msys2-runtime/commit/707f9b218a, https://github.com/git-for-windows/msys2-runtime/commit/c312f3d3c6, https://github.com/git-for-windows/msys2-runtime/commit/99a84e0081, https://github.com/git-for-windows/msys2-runtime/commit/6efd127eeb, https://github.com/git-for-windows/msys2-runtime/commit/4076235378, https://github.com/git-for-windows/msys2-runtime/commit/12461acc8a, https://github.com/git-for-windows/msys2-runtime/commit/aeb2e423d9, https://github.com/git-for-windows/msys2-runtime/commit/9f3e592333, and https://github.com/git-for-windows/msys2-runtime/commit/92a0032d1a: adjust the implicit Windows <-> Unix-y path conversion to accommodate common Git invocations better, i.e. accommodating expectations documented by Git's test suite.

https://github.com/msys2/msys2-runtime/pull/182

**Reply by dscho** [reply]
I checked the boxes.

**Comment by dscho** [comment]
> * [git-for-windows/msys2-runtime@ad711d9aa5](https://github.com/git-for-windows/msys2-runtime/commit/ad711d9aa5), [git-for-windows/msys2-runtime@90d8aad19b](https://github.com/git-for-windows/msys2-runtime/commit/90d8aad19b) and [git-for-windows/msys2-runtime@4935495078](https://github.com/git-for-windows/msys2-runtime/commit/4935495078): offer to use the standard Windows profile directories as home directories (as is expected by Git for Windows). [I contributed two iterations of these patches](https://inbox.sourceware.org/cygwin-patches/0Lg1Tn-1YnzUw0ScN-00pcgi@mail.gmx.com/T/#t) to the Cygwin project already, and simply have to find the time to do the final touch-up. It's on my ever-growing TODO list.

This was [accepted](https://inbox.sourceware.org/cygwin-patches/ZH81nBXKrkD6xC3H@calimero.vinschen.de/) into Cygwin.

> * [git-for-windows/msys2-runtime@65ce633af3](https://github.com/git-for-windows/msys2-runtime/commit/65ce633af3) (with [git-for-windows/msys2-runtime@f5f594c2ff](https://github.com/git-for-windows/msys2-runtime/commit/f5f594c2ff) squashed in): this works around (as well as possible) issues calling an MSYS2 runtime from another MSYS2 runtime (or even the Cygwin runtime); It is not possible to reuse the pseudo terminals, but at least most other operations should work.

This was merged in https://github.com/msys2/msys2-runtime/pull/192 and https://github.com/msys2/msys2-runtime/pull/219; I vaguely remember that Cygwin rejected the idea that this could ever be made to work, so I did not try to contribute this patch to the Cygwin project.

> * [git-for-windows/msys2-runtime@bec3d608e8](https://github.com/git-for-windows/msys2-runtime/commit/bec3d608e8): this is a net improvement: it lets `uname` report the commit from which the MSYS2 runtime was built. This should go into msys2/msys2-runtime; We may eventually even want to contribute this to the Cygwin project, they may like it, too.

I snuck that into https://github.com/msys2/msys2-runtime/pull/219.

**Comment by dscho** [comment]
> * [git-for-windows/msys2-runtime@589fac65ff](https://github.com/git-for-windows/msys2-runtime/commit/589fac65ff): this is a real bug fix for running `find` on a FAT file system and should probably go into MSYS2 as-is

I upstreamed this as https://github.com/msys2/msys2-runtime/pull/103, which caused problems, and the underlying bug was then fixed via https://github.com/msys2/msys2-runtime/pull/117 instead.

**Comment by dscho** [comment]
> * Git for Windows' `mingw-w64-git` installs a `/git-bash.exe` (and other files that are most inappropriate for MSYS2). We will need to split them into their own package, e.g. `mingw-w64-git-for-windows-root`, that only gets built and installed in the Git for Windows project.

I did this as part of git-for-windows/MINGW-packages#172.

**Reply by dscho** [reply]
There are now also git-for-windows/build-extra#663, git-for-windows#174, git-for-windows/git#5971 and #26470 (the last one still in progress), all in support of this here initiative.

**Comment by dscho** [comment]
https://github.com/msys2/MINGW-packages/pull/26470 was probably the biggest rock, and that's done now. With https://github.com/git-for-windows/git-for-windows-automation/pull/153 on its way, I believe that the only remaining bits are the MSYS2 runtime patches that haven't been upstreamed yet.
